contracts/modules/staking/StakingPool.sol

```diff
@@ -1240,18 +1253,84 @@ contract StakingPool is IStakingPool, ERC721 {
     return ownerOf(0);
   }
 
-  /* management */
+  /* pool management */
 
-  function addProducts(ProductParams[] memory params) external onlyManager {
-    totalSupply = totalSupply;  // To silence view fn warning. Remove once implemented
-    params;
+  function setProducts(uint[] calldata productsToUpdate, ProductParams[] memory params) external onlyManager {
+    _setProducts(productsToUpdate, params);
   }
 
-  function removeProducts(uint[] memory productIds) external onlyManager {
-    totalSupply = totalSupply;  // To silence view fn warning. Remove once implemented
-    productIds;
+  function _setProducts(uint[] calldata productsToUpdate, ProductParams[] memory params) internal {
+
+    (uint24 globalCapacityRatio, uint16[] memory capacityReductionRatios) = ICover(coverContract).getCapacityRatios(productsToUpdate);
+
+    uint32 _totalTargetWeight = totalTargetWeight;
+    uint32 _totalEffectiveWeight = totalEffectiveWeight;
+
+    uint numProductsUpdated;
+
+    for (uint i = 0; i < params.length; i++) {
+      ProductParams memory _param = params[i];
+      StakedProduct memory _product = products[_param.productId];
+
+      if (_product.nextPriceUpdateTime == 0) {
+        Product memory coverProduct = ICover(coverContract).products(_param.productId);
+        require(coverProduct.initialPriceRatio > 0, "StakingPool: Product deprecated or not initialized");
+        _product.nextPrice = coverProduct.initialPriceRatio;
+        _product.nextPriceUpdateTime = uint32(block.timestamp);
+      }
+
+      if (_param.setPrice) {
+        require(_param.targetPrice <= TARGET_PRICE_DENOMINATOR, "StakingPool: Target price too high");
+        _product.targetPrice = _param.targetPrice;
+      }
+
+      if (_param.setWeight) {
+          require(_param.targetWeight <= WEIGHT_DENOMINATOR, "StakingPool: Cannot set weight beyond 1");
+          require(numProductsUpdated < productsToUpdate.length, "StakingPool: Must update product to adjust weights");
+          require(productsToUpdate[numProductsUpdated] == _param.productId, "StakingPool: Must update product to adjust weights");
+
+          uint8 previousEffectiveWeight = _product.lastEffectiveWeight;
+          _product.lastEffectiveWeight = _getEffectiveWeight(_param.productId, _param.targetWeight, globalCapacityRatio, capacityReductionRatios[numProductsUpdated++]);
+          _totalEffectiveWeight -= previousEffectiveWeight;
+          _totalEffectiveWeight += _product.lastEffectiveWeight;
```

Comment:
Simpler and clearer to:
`_totalEffectiveWeight = _totalEffectiveWeight - previousEffectiveWeight + _product.lastEffectiveWeight;`